core/src/main/java/com/linecorp/armeria/common/multipart/MultipartDecoder.java

```diff
@@ -98,16 +125,239 @@ public long demand() {
+
+        if (!subscribedUpdater.compareAndSet(this, 0, 1)) {
+            if (executor.inEventLoop()) {
+                abortLateSubscriber(subscriber, abortCause);
+            } else {
+                executor.execute(() -> abortLateSubscriber(subscriber, abortCause));
+            }
+            return;
+        }
+        final MultipartSubscriber delegatedSubscriber = new MultipartSubscriber(subscriber, executor);
+        this.delegatedSubscriber = delegatedSubscriber;
+        decoded.subscribe(delegatedSubscriber, executor, options);
+        subscribedCompleteFuture.complete(null);
+    }
+
+    private static void abortLateSubscriber(Subscriber<? super BodyPart> subscriber,
+                                            @Nullable Throwable throwable) {
+        subscriber.onSubscribe(NoopSubscription.get());
+        if (throwable != null) {
+            subscriber.onError(throwable);
+        } else {
+            subscriber.onError(new IllegalStateException("subscribed by other subscriber already"));
+        }
     }
 
     @Override
     public void abort() {
-        decoded.abort();
+        abort(AbortedStreamException.get());
     }
 
     @Override
     public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        if (subscribed == -1) {
+            // Already aborted before
+            return;
+        }
+        abortCause = cause;
+        if (subscribedUpdater.compareAndSet(this, 0, -1)) {
+            // There is no subscriber
+            return;
+        }
         decoded.abort(cause);
     }
+
+    BodyPartPublisher onBodyPartBegin() {
+        return new BodyPartPublisher();
+    }
+
+    void requestUpstreamForBodyPartData() {
+        @Nullable
+        final MultipartSubscriber delegatedSubscriber = this.delegatedSubscriber;
+        if (delegatedSubscriber != null) {
+            delegatedSubscriber.requestUpstreamForBodyPartData();
+        }
+    }
+
+    /**
+     * Called by delegated MultipartSubscriber.
+     * e.g.
+     * 1. MultipartDecoder#abort -> MultipartSubscriber#onError -> MultipartSubscriber#cleanup ->
+     *    MultipartDecoder#cleanup
+     * 2. MultipartSubscriber#onComplete -> MultipartSubscriber#cleanup -> MultipartDecoder#cleanup
+     */
+    private void cleanup() {
+        if (delegatedSubscriber != null) {
+            delegatedSubscriber = null;
+        }
+    }
+
+    final class BodyPartPublisher extends DefaultStreamMessage<HttpData> {
+        @Override
+        protected void onRequest(long n) {
+            // Because whenConsumed will run in the same thread(called by onRequest) after looping the existing
+            // queue.(onRequest & event notification in whenConsumed will run in the same executor specified
+            // at subscribe)
+            // So if there is no change, it means there is no buffered data processed.
+            whenConsumed().thenRun(() -> {
+                // There isn't any buffered data, or it's not enough
```

Comment:
Question: Is there any chance that we should request more to the upstream?
e.g. The downstream request 1 but the body part was split into 3 frames so we should request 3 to the upstream. 🤔

Reply:
IIUC, if we mean HttpData of body at MimeParser in that case, downstream should request more by self, because what downstream request is the HttpData and it means subscribe needs request more manually. e..g 1GB file is split in many frames, so downstream control the speed 1 by 1.

Reply:
Thanks for the explanation. 😄